### ISSUE TITLE
check conditions for jit cache misses

### ISSUE THREAD

**Issue by pmelchior** [pmelchior/scarlet2#120]
We have reasons to strongly suspect that processes that fit different scenes/observations trigger cache misses for the jitted `Scene._make_step` function. This issue is about diagnosing the problem.

JAX provides a utility function for cache misses: https://docs.jax.dev/en/latest/persistent_compilation_cache.html#examining-cache-misses
Let's find out if merely providing 
1. a different scene (with the same parameterization) but the same observation
2. a different observation with the same scene parameterization
triggers a cache miss.

I suspect it only happens with option 2. If so, we need to investigate how to avoid the cache miss for observations that are fixed-format and only differ by the PSF because that'll be the most common configuration for large-scale survey processing.

**Comment by b-remy** [comment]
There seems to be a cache miss for the `_make_step` function in both cases, resulting in it being recompiled for each new fit. Here is the warning:

```
TRACING CACHE MISS at /home/ben/Documents/repos/scarlet2/scarlet2/scene.py:306 (fit) because:
  for _make_step defined at /home/ben/micromamba/envs/galsim/lib/python3.9/site-packages/equinox/_jit.py:37
explanation unavailable! please open an issue at https://github.com/google/jax
Compiling _make_step with global shapes and types [ShapedArray(float32[2]), ShapedArray(float32[1]), ShapedArray(float32[31,31]), ShapedArray(float32[1,128,128]), ShapedArray(float32[1,128,128]), ShapedArray(int32[]), ShapedArray(float32[1]), ShapedArray(float32[31,31]), ShapedArray(float32[1]), ShapedArray(float32[31,31]), ShapedArray(int32[])]. Argument mapping: [UnspecifiedValue, UnspecifiedValue, UnspecifiedValue, UnspecifiedValue, UnspecifiedValue, UnspecifiedValue, UnspecifiedValue, UnspecifiedValue, UnspecifiedValue, UnspecifiedValue, UnspecifiedValue].
Finished jaxpr to MLIR module conversion jit(_make_step) in 0.058747291564941406 sec
Finished XLA compilation of jit(_make_step) in 0.20924758911132812 sec
100%|████| 200/200 [00:00<00:00, 249.44it/s, loss=15199.57, max_change=0.004965]
```

don't know how to fix this yet, as it is not very explicit

**Comment by pmelchior** [comment]
Is this more than the initial cache miss? For the first fit, there needs to be a compilation. Do we know that this is for the second call to `fit` (the one that either has a different scene or a different observation)?

**Comment by b-remy** [comment]
Yes, this is for the first, second and every other call to `fit`

**Comment by pmelchior** [comment]
So, every call to fit causes a cache miss? Great!

**Comment by pmelchior** [comment]
So, with some testing Benjamin and I realized that it's the `static` attributes that interfere with the cache. The solution for #121 also fixes this issue.

**Comment by pmelchior** [comment]
Hmm. That's the model evaluation itself. On would think that equinox ensures that's done correctly. Let's try a simple test case. 
Something like
```python
class Test(scarlet2.Module):
    a: jnp.array
    def __call__(self):
        return a

@filter_fit
def func(t):
    return t()
```

If you call `func` multiple times, does it create cache misses?

**Comment by pmelchior** [comment]
Further progress: as per [this](https://github.com/patrick-kidger/equinox/issues/1062#issuecomment-3145974755) answer, it's a change of non-array arguments that cause the recompilation. I thought these could be part of our model specification, but it is the optimizer.

The current `_make_step` function has the following signature:
https://github.com/pmelchior/scarlet2/blob/e20552b08e11fa6cf8bc309b2b0bdfec31cc1ad1/src/scarlet2/scene.py#L391-L393

If I define `optim` globally, the cache misses disappear. That's good, but still not ideal because a call to `fit` currently defines the optimizer with a specific stepsize and schedule, so we'd need to ensure that `optim` (and thus `_make_step`) is updated because the full gradient update is jitted. The other option is to limit the jitted function to the gradient computation and keep the update step out of the jitted path. I'll do some experiments with either option.

**Comment by pmelchior** [comment]
So, jitting just the gradient computation but not the update of the model prevents the cache miss but doubles the runtime of the quickstart `fit` call on my CPU, from ~330 it/s to ~150 it/s (and that's including the cache miss for the 330it/s, but excluding it for the 150 it/s, so the compiled run-time costs of this solution are even worse.). That's too expensive. I will therefore stick to the full `_make_step` function and accept the recompilation on repeated runs with the exact same parameterization. This is a little too niche to bother and sacrifice so much performance for.

For future reference: If repeated call with the same parameters/schedule are needed (e.g. for large scale bulk processing), the way to avoid the cache miss and recompilation is to return the optimizer _instance_ by `Scene.fit` and reuse it for future calls the `Scene.fit`.
